xds/src/test/java/io/grpc/xds/LookasideLbTest.java

```diff
@@ -245,80 +185,179 @@ public int getPriority() {
 
       @Override
       public String getPolicyName() {
-        return "supported1";
+        return "round_robin";
       }
 
       @Override
       public LoadBalancer newLoadBalancer(Helper helper) {
-        return mock(LoadBalancer.class);
+        String localityName = helper.getAuthority();
+        childHelpers.put(localityName, helper);
+        LoadBalancer balancer = mock(LoadBalancer.class);
+        childBalancers.put(localityName, balancer);
+        return balancer;
       }
     });
+  }
+
+  @After
+  public void tearDown() {
+    lookasideLb.shutdown();
 
+    for (LoadBalancer childBalancer : childBalancers.values()) {
+      verify(childBalancer).shutdown();
+    }
+
+    if (bootstrapChannel != null) {
+      assertThat(bootstrapChannel.isShutdown()).isTrue();
+    }
+    if (isFullFlow) {
+      xdsClientPoolFromResolveAddresses.returnObject(xdsClientFromResolvedAddresses);
+      assertThat(xdsClientPoolFromResolveAddresses.xdsClient).isNull();
```

Comment:
This is to verify `XdsClientPool.getObject()` and `returnObject()` are used correctly in `LookasideLb`.  If `LookasideLb` forgets to call `returnObject()` in `shutdown()`, this test will break.

Reply:
That's not what it is doing. In line 214, you called `xdsClientPoolFromResolveAddresses.returnObject(...)` and in line 215 you verify `xdsClientPoolFromResolveAddresses.xdsClient` is destroyed. Line 215 verifies line 214, this is not correct.

Reply:
The test calls `getObject()` and `returnObject()` (line 214)  both once, line 215 verify `LookasideLb` calls equal times of `getObject()` and `returnObject()`, otherwise it will break.

Reply:
But, `getObject()` and `returnObject()` are called by your __test implementation__, not __code under test__. That is, you are verifying the correctness of your test code instead of the code running in `LookasideLb.java`.

Reply:
The test implement calls equal times of `getObject()` and `returnObject()` simulating `XdsNameResolver` and `CdsLoadBalancer` calling equal times of `getObject()` and `returnObject()` . That's the precondition to test the code under test (`LookasideLb` calling equal times of `getObject()` and `returnObject()`)

Reply:
Alright, that's not a straightforward test. 

The test strategy for no resource leaking for the whole system should be:
- ensure `LookasideLb` does not leak it (test in this test class).
- ensure `CdsLoadBalancer` does not leak it (test in CDS policy's test).
- ensure `XdsNameResolver` does not leak it (test in the resolver's test).

Given that `RefCountedXdsClientObjectPool`'s implementation is correct (tested it) and the above three are the only places that touch it, the whole system would not have resource leak. Therefore, all you need in `LookasideLb`'s unit test is to ensure it does not leak XdsClient, no matter other components do or not. The most straightforward way to do it is: pass a fake `ObjectPool<XdsClient>` (instead of `RefCountedXdsClientObjectPool`) to `LookasideLb` such that it internally keep tracking of `getObject()` and `returnObject()` calls (e.g., `AtomicBoolean`).

> The test implement calls equal times of getObject() and returnObject() simulating XdsNameResolver and CdsLoadBalancer calling equal times of getObject() and returnObject() .

This doesn't make sense. This only proves: _if `XdsNameResolver` and `CdsLoadBalancer` do not leak, then `LookasideLb` does not leak_. But our goal is: `LookasideLb` does not leak.

Reply:
The goal "`LookasideLb` does not leak" is verified by:

Make sure everything in test implementation does not leak (a precondition), then the whole test does not leak.

Using a fake `ObjectPool<XdsClient>` is a good alternative, the downside of it is that either the fake ObjectPool has to implement shutting down the XdsClient when the resource is fully returned, that's basically a re-implementation RefCountedXdsClientObjectPool(); or `tearDown()` has to manually shutdown the XdsClient which looks wired.

Reply:
Updated to use a fake ObjectPool.